name.abuchen.portfolio/src/name/abuchen/portfolio/online/impl/variableurl/Factory.java

```diff
@@ -28,27 +55,57 @@ public static VariableURL fromString(CharSequence input)
 
         parts.add(input.subSequence(lastEnd, input.length()));
 
-        if (parts.size() > 1)
+        boolean isMacro = false;
+        List<Macro> macros = new LinkedList<>();
+        Macro macro = null;
+
+        for (CharSequence part : parts)
         {
-            try
+            if (isMacro)
             {
-                return new FormattedDate(parts);
-            }
-            catch (IllegalArgumentException ignored)
-            {
-                // try next pattern
-            }
+                for (MacroParser macroParser : MACRO_PARSERS)
+                {
+                    try
+                    {
+                        macro = macroParser.parse(part);
+                    }
+                    catch (IllegalArgumentException e)
+                    {
+                        continue;
+                    }
 
-            try
-            {
-                return new PageNumber(parts);
+                    macros.add(macro);
+                    break;
+                }
+
+                if (macro == null)
+                {
+                    return new ConstURL(Collections.singletonList(new ConstString(input)));
+                }
+
+                macro = null;
             }
-            catch (IllegalArgumentException ignored)
+            else
             {
-                // try next pattern
+                macros.add(new ConstString(part));
             }
+
+            isMacro = !isMacro;
```

Comment:
Die Logik verstehe ich nicht 100%. Im Prinzip könnten auch zwei Macro direkt hintereinander kommen, z.B. als ```something{PAGE}{ISIN}```. Oder das Pattern könnte als normaler Text auftauchen, also ```something{WKN}{ISIN}PAGE```.

Und ist es nicht sowieso einfacher das Macro in der oberen Schleife direkt zuzuordnen - wenn das Pattern gefunden wurde?

Reply:
Stimmt. Wenn man die Schleifen zusammenlegt, kann man schauen, ob der String zwischen `}` und `{` nicht leer ist. Ich schaus mir an.